Unable to use < as part of a password (admin-cli)

Pepo48 opened the pull request:
* escaped angle bracket characters in password

Closes #21951

shawkins commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/cli/admin/KcAdmTest.java:
```diff
@@ -283,6 +283,14 @@ public void testUserLoginWithDefaultConfig() {
+    public void testPasswordWithSpecialCharacter() {
+        KcAdmExec exe = KcAdmExec.execute("config credentials --server " + serverUrl + " --realm test --user special-character@localhost --password password<");
+        Assert.assertEquals("sh: -c: line 0: syntax error near unexpected token `newline'", exe.stderrLines().get(0));
+        KcAdmExec exe1 = KcAdmExec.execute("config credentials --server " + serverUrl + " --realm test --user special-character@localhost --password 'password\\<'");
```
Is the backslash for the shell or due to something in the command framework?  What happens when you use --password 'password<' ?

Pepo48 replied:
To explain the whole context, angle brackets have special meaning,  hence the `syntax error near unexpected token `newline'`.

In standard bash it can be solved by enquoting a value with single quotes. But once you do it, `jboss.aesh` starts to complain about the same. Moreover, admin-cli is still dependant on some [ancient version](https://github.com/keycloak/keycloak/blob/main/pom.xml#L83) (see [jboss.aesh](https://repository.jboss.org/org/jboss/aesh/aesh/) vs. [org.aesh](https://github.com/aeshell/aesh)).

The only (undocumented) workaround I found was to escape it with backslash.

So you're absolutely right, it's due to the framework.

shawkins replied:
Thanks for the clarification.  So this could happen with any of the cli values.  It might be good to say that not only should < be escaped, but if the value contains \< that will need an additional backslash as well.

vmuzikar replied:
+1 @shawkins
`<` could be anywhere, not just in password.

Pepo48 replied:
@shawkins you mean to mention it in the documentation? Because when it comes to the implementation itself, it behaves exactly like you described. Last Friday I added a new UT that demonstrates it. PTAL. Thanks!

Pepo48 commented:
jfyi, some of the test failures seem to be related to the changes (I missed these during the local testing), I already know about them and I'll try to resolve it ASAP.

vmuzikar commented:
@Pepo48 Just wanted to check – is it ready for another review round?

Pepo48 commented:
> @Pepo48 Just wanted to check – is it ready for another review round?

@vmuzikar (cc: @shawkins) I fixed the failures, added a couple of unit tests and applied the function to more arguments. PTAL - I think that besides `noconfig`, `config`, `server` and `external_token` angle brackets can appear everywhere. `realm` though needs to be handled differently - since realm name affect the structure of the endpoints, I escaped the brackets the HTML way. So you can specify the realm name directly as `--realm '%3C%3E'` or:
![image](https://github.com/keycloak/keycloak/assets/12138171/351bd630-105f-4f02-bcb1-2bac45a3c3f9)

Pepo48 commented:
@shawkins thanks for the suggestion. With the proposed above I currently get the following result:
```
./kcadm.sh config credentials --server http://localhost:8080 --realm master --user \<\> --password \<\>
Logging into http://localhost:8080 as user <> of realm master
```
I guess it's a nicer approach in the end. The realm argument must be handled differently though (because of what I mentioned in my previous comment). I'll adjust it tomorrow and revisit the the corresponding tests.

Thanks again.

vmuzikar commented:
@Pepo48 @shawkins Is this ready  for another review round?

(@shawkins Feel free to merge when ready.)

Pepo48 commented:
@vmuzikar it's not ready yet.

Pepo48 committed:
Unable to use < as part of a password (admin-cli)

* escaped angle bracket characters in password

Closes #21951

Signed-off-by: Peter Zaoral <pzaoral@redhat.com>

Pepo48 commented on integration/client-cli/admin-cli/src/main/java/org/keycloak/client/admin/cli/KcAdmMain.java:
```diff
@@ -79,24 +79,23 @@ public static void main(String [] args) {
             arguments.addAll(Arrays.asList(args));
+            int realmIndex = arguments.indexOf("--realm");
+            if (realmIndex != -1 && realmIndex < arguments.size() - 1) {
+                arguments.set(realmIndex + 1, arguments.get(realmIndex + 1).replace("<", "%3C").replace(">", "%3E"));
```
Afaik, we are not doing the explicit conversion anywhere else, in server terms we leave it to Quarkus (specifically UriHandler in rest-easy-reactive-common) to handle (and transform) it, see:
![image](https://github.com/keycloak/keycloak/assets/12138171/fa12dbf8-0cff-4bde-a1ca-842e38c015af)

This it the [place](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/admin/RealmsAdminResource.java#L147) where we get the escaped value which is then used in the well-known endpoints within the administration console.

Currently, if the raw angle brackets are left in, it causes the following error:
```
Logging into http://localhost:8080 as user <> of realm <>
Illegal character in path at index 29: http://localhost:8080/realms/<>/protocol/openid-connect/token
Try 'kcadm.sh help config credentials' for more information
```

That's the reason I transformed it.

shawkins replied:
> Currently, if the raw angle brackets are left in, it causes the following error:

If I understand correctly the path method here:
```AdminRoot.realmsUrl(session.getContext().getUri()).path(realm.getName()).build()```
is generally handling reserved characters.  However <> since they aren't strictly forbidden, they are still allowed.  That is different than the behavior you are showing above with the UriHandler, so I'd say this should be a bug in resteasy, or whatever is backing the implenentation of the UriBuilder path method because <> should not be allowed in path segments, only the query portion.

If you url encode the <> won't the path method just enode that further?  % -> %3C -> %253C in which case it still won't be interpretted correctly.

Pepo48 replied:
> If you url encode the <> won't the path method just enode that further?
No, here I created a realm named %3C and this is the result:
![image](https://github.com/keycloak/keycloak/assets/12138171/b8cf2542-9f3d-4ce6-b864-afc3f53b9317)

The difference in processing can be spotted in this method [1]. So that shouldn't be a problem.

But there definitely is a bug and I spotted it once I tried to create a realm named %3C. The console allows you to create it but when you try to access Realm settings, this happens (notice that the value is transformed to < in the url):
![image](https://github.com/keycloak/keycloak/assets/12138171/bb1d585b-c3bb-482b-b678-fe3ccd805a91)

I went further and created a realm named < right afterwards. Then I switched back to the previous %3C realm and tried to access the setting again. The error is gone but only because I got redirected from %3C to < realm immediately after I click on Realm settings.

I think that this certainly deserves a follow-up, since I think it already went beyond the scope of this PR.

cc: @vmuzikar

[1] https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/Encode.java#L398

vmuzikar commented on docs/documentation/server_admin/topics/admin-cli.adoc:
```diff
@@ -140,6 +140,12 @@ For example, when performing an operation, specify all the information required
 $ kcadm.sh get realms --no-config --server http://localhost:8080{kc_base_path} --realm master --user admin --password admin
 ----
 
+In case you want to use the angle bracket character (`<` or `>`) in an argument value, you need to escape every angle bracket character with the backslash:
```
This is actually related to https://github.com/keycloak/keycloak/pull/25218.

I think we can be a bit more explicit here and in a generic way (it's not just about angle brackets but also e.g. for dollar sign) describe the escaping for double and no quotes and that it's not needed for single quotes, just to make it crystal clear for readers. I know it's basic shell functionality but still might save some time for somebody.

vmuzikar reviewed (commented):
> Speaking about the scope of this PR, will it be ok, if I completely leave the [realm transformation](https://github.com/keycloak/keycloak/blob/c5b3c2640868782195c26fda1a5c2468192d5edf/integration/client-cli/admin-cli/src/main/java/org/keycloak/client/admin/cli/KcAdmMain.java#L82L85) out (and as discussed above the documentation bit as well)?

Yes, please. No special handling of realm should be needed. Any resource ID (which is just a string that can contain special characters) can appear in the URL, so it's a generic problem of how we build the URLs in the admin client rather than a realm name specific issue.

shawkins commented:
> This is actually related to https://github.com/keycloak/keycloak/pull/25218.

Just to make sure, you're talking about the bit at the end where examples are given for how to treat \ and $ metacharacters in a bash shell or properties file.  The additional config related escaping for $ or $$ doesn't apply here.

Pepo48 commented:
> This is actually related to https://github.com/keycloak/keycloak/pull/25218.

I removed the docs change from this PR as per discussion above.

shawkins reviewed (approved):
LGTM, thanks @Pepo48

vmuzikar commented:
> Just to make sure, you're talking about the bit at the end where examples are given for how to treat \ and $ metacharacters in a bash shell or properties file. The additional config related escaping for $ or $$ doesn't apply here.

Yup, sorry, should've clarified better.

> I removed the docs change from this PR as per discussion above.

I think we could have that docs change just in a more generic way. But yeah, it's not specific to Keycloak, we'd be basically explaining basic shell functionality. So removing works for me.
